lib/graphql/preload/loader.rb

```diff
@@ -34,10 +35,13 @@ def perform(records)
       end
 
       private def preload_association(records)
+        if scope && (scope.klass != model.reflect_on_association(association).klass)
+          self.scope = nil
+        end
         if ActiveRecord::VERSION::MAJOR > 3
-          ActiveRecord::Associations::Preloader.new.preload(records, association)
+          ActiveRecord::Associations::Preloader.new.preload(records, association, scope)
         else
-          ActiveRecord::Associations::Preloader.new(records, association).run
+          ActiveRecord::Associations::Preloader.new(records, association, scope).run
```

Comment:
Hey @CyborgMaster! Thanks for the great PR!

I was thinking about how you pass the scope all the way down here, and realized that in Rails 3, `ActiveRecord::Associations::Preloader#new` actually takes an `options` hash as the third argument. It looks like if we want to preserve Rails 3 compatibility, we may have to modify the interface for this gem to accept a hash of conditions instead of a callable.

I took a shot at implementing that here: https://github.com/ConsultingMD/graphql-preload/compare/feature/preload-with-scope?expand=1. Would you mind taking a look and letting me know what you think? Does this also solve your problem (and does it present an interface you're happy with)?

Reply:
Right, I didn't look at Rails 3 compatibility at all, so we'll have to address that somehow.  

Just to clarify, in Rails 4, the third parameter isn't a callable, it takes a Relation, something like `Transactions.where('posted_at > 2017-10-01')`.

The problem I am trying to work around is the preload scope needs both the args and context, which is why it got passed all the way down in that strange manner in my original PR.  

In the example I gave above the preload scope would end up being something like this (simplified in this example to just the pertinent code):

``` ruby
preload_scope lambda { |args, ctx|
  Transactions
    .where(source_id: ctx[:source_id])
    .where('posted_at > ?', args[:filter][:posted_at_gt])
}
```

so we need the context for security purposes and the args to determine the filters applied at that level, neither of which we have if we just do an instance eval.  

Thoughts?

Reply:
I see what you mean. I was thinking about this a bit over the weekend; we could modify the interface to return the `conditions` expected by Rails 3:

```ruby
preload_conditions ->(obj, args, ctx) do
  ['source_id = ? and posted_at > ?', ctx[:source_id], args[:filter][:posted_at_gt]]
end
```

Then, pass that straight to Rails 3, or turn it into a relation by calling `klass.where(preload_conditions)` in Rails 4. Of course, that's not a great interface, and more importantly, it sacrifices a lot of functionality (like calling a named scope). It might be easier just to cut the Rails 3 compatibility for this feature.

Reply:
Yeah, or doing any other advanced things in, say Arel, that are cumbersome to do in String form (this is actually what we are doing in our system.  I'd be fine if we just had it work for Rails 4.